### DISCUSSION TITLE
timer precision in ms… why not µs?

### DISCUSSION THREAD

**Discussion by zx80** [libuv/libuv#4122] [Q&A]
For throttling operations in a benchmark, I would need  µs precision.

For instance on a local socket on my laptop postgres can reach over 90000 tps on one thread for simple selects, thus a benchmark can be throttled say to 50000 tps, that is one transaction is scheduled every 20 µs on average:

```shell
pgbench -S -T 3 -P 1 -M prepared -R 50000
pgbench (15.4 (Ubuntu 15.4-1.pgdg22.04+1))
starting vacuum...end.
progress: 1.0 s, 49350.5 tps, lat 0.286 ms stddev 1.223, 0 failed, lag 0.274 ms
progress: 2.0 s, 50158.4 tps, lat 0.089 ms stddev 0.130, 0 failed, lag 0.078 ms
progress: 3.0 s, 49984.7 tps, lat 0.134 ms stddev 0.289, 0 failed, lag 0.122 ms
number of clients: 1
number of threads: 1
latency average = 0.169 ms
latency stddev = 0.731 ms
rate limit schedule lag: avg 0.157 (max 8.849) ms
tps = 49941.354136 (without initial connection time)
```

Ok, I understand that an event loop would not actually be precise enough to  start something at 1 µs around the time set, but even 10 or 50 µs would be significantly better than the offered *ms* precision.

Is there a way to improve the precision offered by `libuv` so that the thing above would be possible? Am I missing something?
The documentation advertises *High resolution clock*, but ms resolution does not stuck me as high resolution on a GHz multi-core processor.

**Comment by vtjnash** [comment]
This might be a good v2 feature?

**Reply by zx80** [reply]
I gather from your reply that currently *ms* is all that can be achieved from the API.

This is definitely a desirable feature if `libuv` would enable the use case I outline, otherwise it looks like a no go.

**Comment by zx80** [comment]
While I'm at it, the *current time* precision should also be better than *ms*  provided by `uv_now` if this thing is to be improved.

**Comment by bnoordhuis** [comment]
As a bit of historic background: timers are mainly intended as companions for things like TCP sockets, where you want operations to time out eventually but you can tolerate some imprecision. Not all platforms libuv runs on have millisecond precision, let alone sub-ms precision.

As a practical concern, to get microsecond precision on Linux, libuv would need to add a timerfd to the event loop and then repeatedly arm it.

For the common case, that's probably slower because of the additional system calls (although we might be able to elide those when millisecond precision is good enough) and it's dependent on the kernel providing a timer source with good enough granularity.

It's probably not a problem for desktop systems but systems optimized for batch processing sometimes run at only 250 Hz.

There's also Linux's timer slack feature to consider. You can opt out of it but a library like libuv probably shouldn't make such thread/process-wide changes.

**Reply by zx80** [reply]
ISTM that `libuv` should give access to the available underlying precision, whatever it is, not decide to truncate.

I understand that the initial *view of the world* of `libuv` is abstracting TCP/IP connections, but event loops have a more general use case.

I'm not sure about the `timerfd` comment in Linux. ISTM that `pgbench` works with *µs* without doing such a thing, it just uses *ppoll* or *select* native timeouts.

**Reply by bnoordhuis** [reply]
Milliseconds are the underlying precision because that's what the epoll_pwait system call uses. Linux 5.11 has a epoll_pwait2 syscall that takes a `struct timespec` but libuv supports older kernels so milliseconds it is.

Libuv was initially written with support for linux 2.6.9, hence timerfds (added in 2.6.25) wouldn't have been an option either for the same reason.

> pgbench works with µs without doing such a thing, it just uses ppoll or select native timeouts.

select and (p)poll don't scale. pgbench polls a few hundred file descriptors at most, libuv routinely handles hundreds of thousands.

Having said all that, there's nothing stopping you from creating a timerfd yourself and wrapping it in a `uv_poll_t`. And, technically speaking, there's no reason libuv couldn't grow a new `uv_hrtimer_t` API that uses the highest-precision time source available, but:

1. I'm fairly sure 99.99% of libuv users don't need that kind of precision (because this discussion never really came up before), and

2. The difference in precision across platforms probably makes it rather useless for people writing cross-platform code (which is why most people use libuv in the first place)

If someone wants to work on that, by all means go ahead, just don't expect any libuv maintainers to do it for you (for free anyway, some of us may be open to contracting.)
